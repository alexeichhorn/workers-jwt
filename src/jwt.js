import { getEncodedMessage, getDERfromPEM, str2ab, b64encode } from './utils';
import './globalThis';

const algorithms = {
  RS256: {
    name: 'RSASSA-PKCS1-v1_5',
    hash: { name: 'SHA-256' },
  },
};

export const getHeader = (alg, headerAdditions) => ({
  ...headerAdditions,
  alg,
  typ: 'JWT',
});

// XXX https://developers.google.com/identity/protocols/OAuth2ServiceAccount#jwt-auth
export const getToken = async ({
  privateKeyPEM,
  payload,
  alg = 'RS256',
  headerAdditions = {},
  cryptoImpl = null,
}) => {
  const algorithm = algorithms[alg];
  if (!algorithm) {
    throw new Error(`@sagi.io/cfw-jwt: Unsupported algorithm ${alg}.`);
  }

  if (!globalThis.crypto) {
    if (!cryptoImpl) {
      throw new Error(`@sagi.io/cfw-jwt: No crypto nor cryptoImpl were found.`);
    }
    globalThis.crypto = cryptoImpl;
  }

  const privateKeyDER = getDERfromPEM(privateKeyPEM);
  const privateKey = await crypto.subtle.importKey(
    'pkcs8',
    privateKeyDER,
    algorithm,
    false,
    ['sign']
  );

  const header = getHeader(alg, headerAdditions);
  const encodedMessage = getEncodedMessage(header, payload);
  const encodedMessageArrBuf = str2ab(encodedMessage);

  const signatureArrBuf = await crypto.subtle.sign(
    algorithms.RS256.name,
    privateKey,
    encodedMessageArrBuf
  );

  const encodedSignature = b64encode(signatureArrBuf);
  const token = `${encodedMessage}.${encodedSignature}`;
  return token;
};

// Service Account Authoriazation without OAuth2:
// https://developers.google.com/identity/protocols/OAuth2ServiceAccount#jwt-auth
// Service Account Auth for OAuth2 Tokens: Choose "HTTP / REST" for:
// https://developers.google.com/identity/protocols/OAuth2ServiceAccount
export const getTokenFromGCPServiceAccount = async ({
  serviceAccountJSON,
  aud,
  scope = '',
  alg = 'RS256',
  expiredAfter = 3600,
  cryptoImpl = null,
}) => {
  const {
    client_email: clientEmail,
    private_key_id: privateKeyId,
    private_key: privateKeyPEM,
  } = serviceAccountJSON;

  const headerAdditions = { kid: privateKeyId };
  const header = getHeader(alg, headerAdditions);

  const iat = parseInt(Date.now() / 1000);
  const exp = iat + expiredAfter;
  const iss = clientEmail;
  const sub = clientEmail;
  const payload = { aud, iss, sub, iat, exp };

  !!scope && Object.assign(payload, { scope });

  return getToken({ privateKeyPEM, payload, alg, headerAdditions, cryptoImpl });
};
