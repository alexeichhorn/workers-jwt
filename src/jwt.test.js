import fs from 'fs';
import * as jwt from './jwt';

describe('cfw-jwt', () => {
  const WebCrypto = require('node-webcrypto-ossl');
  const crypto = new WebCrypto();

  global.Date.now = jest.fn(() => 1530518207007);

  test('getToken', async () => {
    // XXX Don't worry - this service account was deleted (i.e. can't be abused).
    const sa = require('./testdata/service_account.json');
    const iat = parseInt(Date.now() / 1000);
    const exp = iat + 60 * 60;
    const iss = sa.client_email;
    const sub = sa.client_email;
    const aud = 'https://sagi.io';
    const scope = 'bla:xyz';
    const payload = { iss, sub, iat, exp, aud, scope };

    const headerAdditions = { kid: sa.private_key_id };
    const privateKeyPEM = sa.private_key;
    const token = await jwt.getToken({
      privateKeyPEM,
      payload,
      headerAdditions,
      cryptoImpl: crypto,
    });
    expect(token).toMatchSnapshot();
  });

  test('getTokenFromGCPServiceAccount', async () => {
    // XXX Don't worry - this service account was deleted (i.e. can't be abused).
    const serviceAccountJSON = require('./testdata/service_account.json');

    const aud = 'https://sagi.io';
    const scope = 'bla:xyz';

    const token = await jwt.getTokenFromGCPServiceAccount({
      serviceAccountJSON,
      aud,
      scope,
      cryptoImpl: crypto,
    });
    expect(token).toMatchSnapshot();
  });
});
