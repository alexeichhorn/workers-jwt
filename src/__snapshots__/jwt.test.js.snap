// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cfw-jwt getTokenFromGCPServiceAccount 1`] = `"eyJraWQiOiI2ODk1ZWQ1MjdlNDI3ODYxMmM2ODNlMzFlZTIwYzEwZWZjZTE0ODUxIiwiYWxnIjoiUlMyNTYiLCJ0eXAiOiJKV1QifQ.eyJhdWQiOiJodHRwczovL3NhZ2kuaW8iLCJpc3MiOiJjZnctcmVzdC1hcGktdGVzdDJAYXhhbXBsZS1wcm9qZWN0LmlhbS5nc2VydmljZWFjY291bnQuY29tIiwic3ViIjoiY2Z3LXJlc3QtYXBpLXRlc3QyQGF4YW1wbGUtcHJvamVjdC5pYW0uZ3NlcnZpY2VhY2NvdW50LmNvbSIsImlhdCI6MTUzMDUxODIwNywiZXhwIjoxNTMwNTIxODA3LCJzY29wZSI6ImJsYTp4eXoifQ.QAlhoyzrTw1s087-CwohPgdZ9KSFckeCkQJU5sVN1oA0AMKoHXopk8DQdyyqGSBSrEGfKz127Kdc4BahMpSvsOuA6rpCiT7KxOv4skPm8fVGRA2Gw6hcDWpJTBmk7XZAXTxjIHR1YGbv3g9KNC6_d4aTvr3_mOJTjlDFsvIGtuHKMGDfDxHLk3ctE1V911MfouCyJdCS73-abRyvHETLcRREg11vWUWy-5ELR2kLbfsvBcQOfthV2Oqv94kJaxeE9kV5uqR-SR-gGSmzWqFQXScHXBSDZ-3-7FYPdBEHk1LcCq71iHbmTLdl5EgJCc88N381Ad5QTpVCvXt45uA8Rg"`;
